text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,395 @@
+  [private dependencies](https://github.com/rust-lang/cargo/issues/2064), which
+  should have a separate RFC in the near future.
+
+- Once they're satisfied with the change to `xml-rs` (and any other intermediate
+  crates), they'll need to make PRs and request a new publication to
+  crates.io. But they would like to cleanly continue local development in the
+  meantime, with an easy migration as each PR lands and each crate is published.
+
+## The Goldilocks problem
+
+It's likely the a couple of Cargo's existing features have already come to mind
+as potential solutions to the challenges above. But the existing features suffer
+from a Goldilocks problem:
+
+- You might reach for git (or even path) dependencies. That would mean, for
+  example, switching an `xml-rs` dependency in your crate graph from crates.io
+  to point at, for example, a forked copy on github. The problem is that **this
+  approach does not provide enough dependency unification**: if other parts of
+  the crate graph refer to the crates.io version of `xml-rs`, it is treated as
+  an entirely separate library and thus compiled separately. That in turn means
+  that two crates in the graph using these distinct versions won't be able to
+  talk to each other about `xml-rs` types (even when those types are identical).
+
+- You might think that `[replace]` was designed precisely for the use case
+  above. But **it provides too much dependency unification**: it reroutes *all*
+  uses of a particular existing crate version to new source for the crate, even
+  if there are breaking changes involved. The feature is designed for surgical
+  patching of specific dependency versions.
+
+Prepublication dependencies add another tool to this arsenal, with just the
+right amount of dependency unification: the precise amount you'd get after
+publication to crates.io.
+
+# Detailed design
+[design]: #detailed-design
+
+The design itself is relatively straightforward. The Cargo.toml file will
+support a new section for augmenting a source of crates:
+
+```toml
+[augment.crates-io]
+xml-rs = { path = "path/to/fork" }
+```
+
+The listed dependencies have the same syntax as the normal `[dependencies]`
+section, but they must all come form a different source than the source being
+augmented. For example you can't augment crates.io with other crates from
+crates.io! Cargo will load the crates and extract the version information for
+each dependency's name, supplementing the source specified with the version it
+finds. If the same name/version pair *already* exists in the source being
+augmented, then this will act just like `[replace]`, replacing its source with
+the one specified in the `[augment]` section.
+
+Like `[replace]`, the `[augment]` section is only taken into account for the
+root crate (or workspace root); allowing it to accumulate anywhere in the crate
+dependency graph creates intractable problems for dependency resolution.
```

Comment:
I'm not sure what this paragraph means exactly-- does it mean that a dependent crate's `[augment]` section will be ignored? So in the servo/foo/xml-rs example, when resolving deps for servo, cargo ignores the `[augment]` section in foo's Cargo.toml?